src/include/platform/DeviceInstanceInfoProvider.h

```diff
@@ -0,0 +1,124 @@
+    virtual ~DeviceInstanceInfoProvider() = default;
+
+    /**
+     * @brief Obtain the Serial Number from the device's factory data.
+     *
+     * The SerialNumber attribute specifies a human readable serial number
+     *
+     * @param[in, out] buf Buffer to copy string.
+     *                 On CHIP_NO_ERROR return from this function this buffer will be null-terminated.
+     *                 On error CHIP_ERROR_BUFFER_TOO_SMALL there is no guarantee that buffer will be null-terminated.
+     * @param[in] bufSize Size of data, including the null terminator, that can be written to buf.
+     *                    This size should be +1 higher than maximum possible string.
+     *                    This value is valid if the return value is either CHIP_NO_ERROR or CHIP_ERROR_BUFFER_TOO_SMALL.
+     * @returns CHIP_NO_ERROR on success, or another CHIP_ERROR from the underlying implementation
+     *          if access fails.
+     */
+    virtual CHIP_ERROR GetSerialNumber(char * buf, size_t bufSize) = 0;
+
+    /**
+     * @brief Obtain a manufacturing date from the device's factory data.
+     *
+     * The ManufacturingDate attribute specifies the date that the Node was manufactured.
+     *
+     *
+     * @param[out] year Reference to location where manufacturing year will be stored
+     * @param[out] month Reference to location where manufacturing month will be stored
+     * @param[out] day Reference to location where manufacturing day will be stored
+     * @returns CHIP_NO_ERROR on success, or another CHIP_ERROR from the underlying implementation
+     *          if access fails.
+     */
+    virtual CHIP_ERROR GetManufacturingDate(uint16_t & year, uint8_t & month, uint8_t & day) = 0;
+
+    /**
+     * @brief Obtain a Hardware Version from the device's factory data.
+     *
+     * @param[out] hardwareVersion Reference to location where the hardware version integer will be copied
+     * @returns CHIP_NO_ERROR on success, or another CHIP_ERROR from the underlying implementation
+     *          if access fails.
+     */
+    virtual CHIP_ERROR GetHardwareVersion(uint16_t & hardwareVersion) = 0;
+
+    /**
+     * @brief Obtain a Hardware Version String from the device's factory data.
+     *
+     * The HardwareVersionString can be used to provide a more user-friendly value than that
+     * represented by the HardwareVersion attribute.
+     *
+     * @param[in, out] buf Buffer to copy string.
+     *                     On CHIP_NO_ERROR return from this function this buffer will be null-terminated.
+     *                     On error CHIP_ERROR_BUFFER_TOO_SMALL there is no guarantee that buffer will be null-terminated.
+     * @param[in] bufSize Size of data, including the null terminator, that can be written to buf.
+     *                    This size should be +1 higher than maximum possible string.
+     *                    This value is valid if the return value is either CHIP_NO_ERROR or CHIP_ERROR_BUFFER_TOO_SMALL.
```

Comment:
Again, what does it mean for it to be "valid" if the return is CHIP_ERROR_BUFFER_TOO_SMALL?